lib/stdlib/src/erl_expand_records.erl

```diff
@@ -262,6 +270,42 @@ not_a_tuple({op,_,_,_}) -> true;
 not_a_tuple({op,_,_,_,_}) -> true;
 not_a_tuple(_) -> false.
 
+traverse_af(AF, Fun) ->
+    traverse_af(AF, Fun, []).
+traverse_af(AF, Fun, Acc) when is_list(AF) ->
+    [traverse_af(Ast, Fun, Fun(Ast,Acc)) || Ast <- AF];
+traverse_af(AF, Fun, Acc) when is_tuple(AF) ->
+    %% Iterate each tuple element, if the element is an AF, traverse it
+    [[(fun  (List) when is_list(List) ->
+                            traverse_af(List, Fun, Acc);
+            (Tuple) when is_tuple(Tuple)->
+                            case erl_anno:is_anno(Tuple) of
+                                true -> [];
+                                false -> traverse_af(Tuple, Fun, Fun(Tuple,Acc))
+                            end;
+            (_) -> []
+        end)(Term) || Term <- tuple_to_list(AF)],Acc];
+traverse_af(_, _, Acc) -> Acc.
+save_vars({var, _, Var}, _) -> Var;
+save_vars(_, Acc) -> Acc.
+free_variables(AF, Acc) ->
+    try
+            _=traverse_af(AF, fun free_variables1/2, Acc),
+            false
+    catch
+            throw:{error,unsafe_variable} -> true
+    end.
+free_variables1({'fun',_anno,{clauses, _}}, Acc) ->
+    {function,Acc}; %% tag that we are in a 'fun' now that can define new variables
```

Comment:
I don't think this is correct. Only variables defined in the function head will be shadowed. Variables defined in the function body will be matched against variables having the same name in the enclosing function.

In my suggested simplified function, I don't try to handle funs; I think they are used to infrequently to be worth the effort.